File: backend/app/services/audit.py
"""Administrative audit trail.

Any action that changes a complaint's fate - status, priority, assignment,
duplicate merge, rejection - is recorded with the actor, the timestamp and the
old and new values. In a municipal system this is not optional: it is how a
decision gets defended six months later.

Entries are append-only. Nothing in the API updates or deletes an audit row.
"""

from __future__ import annotations

import uuid
from typing import Any

from sqlalchemy.orm import Session

from app.models.system import AuditLog
from app.models.user import User
from app.repositories.system import AuditRepository


class AuditService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.repository = AuditRepository(db)

    def record(
        self,
        *,
        actor: User | None,
        action: str,
        entity_type: str,
        entity_id: uuid.UUID | None = None,
        complaint_id: uuid.UUID | None = None,
        old_value: dict[str, Any] | None = None,
        new_value: dict[str, Any] | None = None,
        note: str | None = None,
        ip_address: str | None = None,
    ) -> AuditLog:
        entry = AuditLog(
            actor_id=actor.id if actor else None,
            # Denormalised so the trail stays readable even if the account is
            # later deleted.
            actor_email=actor.email if actor else None,
            action=action,
            entity_type=entity_type,
            entity_id=entity_id,
            complaint_id=complaint_id,
            old_value=old_value,
            new_value=new_value,
            note=note,
            ip_address=ip_address,
        )
        return self.repository.add(entry)

    def list_recent(self, *, complaint_id: uuid.UUID | None = None, limit: int = 100) -> list[AuditLog]:
        return self.repository.list_recent(complaint_id=complaint_id, limit=limit)

File: supabase/functions/api/services/audit.ts
// Ported from backend/app/services/audit.py. Append-only: nothing here ever
// updates or deletes a row. Always the service client - audit_logs has no
// client-facing insert policy (RPC/system-write only, see the RLS
// migration), and every caller has already passed requireAdmin(req).

import type { SupabaseClient } from "@supabase/supabase-js";
import type { AuthedProfile } from "../_shared/auth.ts";

export interface AuditEntryInput {
  actor: AuthedProfile | null;
  action: string;
  entityType: string;
  entityId?: string | null;
  complaintId?: string | null;
  oldValue?: Record<string, unknown> | null;
  newValue?: Record<string, unknown> | null;
  note?: string | null;
  ipAddress?: string | null;
}

export async function recordAudit(client: SupabaseClient, input: AuditEntryInput): Promise<void> {
  const { error } = await client.from("audit_logs").insert({
    actor_id: input.actor?.id ?? null,
    // Denormalised so the trail stays readable even if the account is later deleted.
    actor_email: input.actor?.email ?? null,
    action: input.action,
    entity_type: input.entityType,
    entity_id: input.entityId ?? null,
    complaint_id: input.complaintId ?? null,
    old_value: input.oldValue ?? null,
    new_value: input.newValue ?? null,
    note: input.note ?? null,
    ip_address: input.ipAddress ?? null,
  });
  if (error) throw error;
}

export async function listRecentAudit(
  client: SupabaseClient,
  opts: { complaintId?: string; limit?: number } = {},
): Promise<Record<string, unknown>[]> {
  let query = client.from("audit_logs").select("*").order("created_at", { ascending: false }).limit(opts.limit ?? 100);
  if (opts.complaintId) query = query.eq("complaint_id", opts.complaintId);
  const { data, error } = await query;
  if (error) throw error;
  return data ?? [];
}
